### PR TITLE
Automatically calculate monster difficulty

### LINKED ISSUE #27308: Monster Difficulty Project: Zombies
NPCs use monster difficulty ratings to determine how much danger they are, which enemies to prioritize, and whether to warn the player about visible monsters.  Unfortunately, monster difficulty ratings are arbitarily defined in JSON, and may not represent actual monster difficulty.

This is a proposal for calculating monster difficulty, using a formula based on melee skill, damage, HP, special attacks, etc.  For the first pass, I want to focus on zombies.

This is spreadsheet showing monster HP, melee skill, maximum melee damage, dodge skill, sum of cut and bash armor values, number of special attacks, number of emitted fields, HP, speed, attack cost, morale, aggression, and day and night vision.

https://docs.google.com/spreadsheets/d/17dM0GQMVNJCsB_AInjb2xlYtb-rKyfMWoyZRRqRvVD4/edit?usp=sharing
I've marked some monsters with my feeling on their rough difficulty (high, medium, low, or nothing special) and noted those monsters that have calculated difficulty numbers lower than their rough difficulty would suggest.

Current estimate of difficult is a multistep process:
* Offensive Base = ( 1 + melee skill ) * maximum melee damage / 25
* Defense Base = ( 1 + dodge skill ) * ( 2 + sum of bash and cut armor value ) / 25
* Special Base = 1 * number of special attacks + 15 * number of emitted fields + fudge factor
* Morale Modifier = Aggression / 10 + Morale / 1000
* Speed Modifier = ( Moves - Attack Cost ) / 100
* HP Modifier = HP / 100
* Vision Modifier = ( Day Vision + 2 * Night Vision ) / 100
* Fudge factor is 2 for monsters with melee range counterattacks like skitterbots, 5 for short range and low accuracy ranged attacks like acid spit, 10 for dangerous ranged attacks like standard turret's SMG, 20 for very dangerous ranged attacks like a rifle turret or zombie necromaner's special abilities, 30 for fatal even with armor attacks like an anti-material turret
Difficulty = ( Offensive Base + Defense Base + Special Base ) * ( Morale Modifier + Speed Modifier + HP Modifier + Vision Modifier )

The logic being a monster is more difficult as it's melee skill and maximum damage increase, as it is harder to hit and damage, and as it has more special abilities.  More HP, effective speed, aggression and morale, and vision have a multiplicative effect on a dangerous monster, but don't directly contribute a monster's difficulty: a monster with no armor or dodge ability, low melee skill and damage, and no special abilities isn't much of threat, even if it has 10,000 HP.

Suggestions on monsters that have inappropriate rough difficulties, or modifications to the formula to improve the estimate, are appreciated.

### PR DESCRIPTION
**Pull request by mlangsdorf**
#### Summary
```SUMMARY: Interface "Automatically calculate monster difficulty"```

#### Purpose of change
Monster difficulty numbers in JSON are fairly arbitrarily assigned, and have a lot of nonsensical values such as baby birds being more difficult than acid zombies, and acid zombies being only slightly more difficult than normal zombies, while zombie kinderlings are nearly as difficult as zombie brutes.  Trusting people to assign sensible difficulty numbers in the absence of solid guidelines has not worked and needs to be abandoned.

Fixes #27308
Instead of taking arbitrary monster difficulty values from JSON, calculate monster difficulty from expected melee damage, defenses, special attacks, hit points, etc, and use the "diff" field to fudge complicated special attacks that don't show up in the normal calculation.

Add a little variation in what monsters that NPCs will warn about.

Display a rough categorization of difficulty when examining a monster, or the precise value in the monster's info when debug_mode is on.

#### Describe the solution
Monster difficulty is calculated from 4 primary factors and then adjusted by a modifier, using these formulas:
difficulty = mod * ( offensive base + defensive base + special base + difficulty from JSON )
* offensive base = ( 1 + melee skill ) * melee_dice * ( melee dice sides + bonus cut damage ) / 25
* defensive base = ( 1 + dodge skill ) * ( 3 + bash armor + cut armor ) / 25
* special base = number of special attacks + 8 * number of emitted fields
* difficulty from JSON = 0 to 40, and represents the danger of special ranged attacks, special defenses like acid spray, or special abilities like a zombie necromancer's revival.  Guidelines are explicit in MONSTERS.md and there are multiple examples for every rating.
* mod = HP / 100 + ( Speed - Attack Cost ) / 100 + ( Morale + Aggression ) / 1000 + ( Day Vision + 2 * Night Vision ) / 100

#### Describe alternatives you've considered
NPC threat evaluation needs to be updated as per #25687, but that turned out to be fairly hard, while this is a quick fix.

### CODE REVIEW

**Comment by Inglonias** [comment]
I don't suppose there's a way for you to display this value in-game to the player, is there? Since NPCs get to know about it?

**Comment by mlangsdorf** [comment]
Not for this PR, but I could add it to the extended description or something.

**Comment by nexusmrsep** [comment]
What's the point of reveling exact number to the player? It would take away the joy of discovering it yourself. If it's feasible to do so however, I'd recommend to textify it into groups from "I eat those monsters for breakfast" to "This monster easts you for breakfast", or along the lines.

**Comment by mlangsdorf** [comment]
We already show rough difficulty by color coding in the monster radar.  I'm thinking just including the rough level in the extended description:
2 < No threat
10 < Mild threat
20 < Threat
30 < Severe threat
30 >= Extreme threat

with an option for debug mode to show the exact number, as well as showing it in the debug menu's Create Monster listing.

**Comment by mlangsdorf** [comment]
Went ahead and added the monster difficulty display.

File: src/monster.cpp [hunk 1]
```diff
@@ -545,6 +545,10 @@ int monster::print_info( const catacurses::window &w, int vStart, int vLines, in
     const auto att = get_attitude();
     wprintz( w, att.second, att.first );
 
+    if( debug_mode ) {
+        wprintz( w, c_light_gray, _( "  Difficulty " ) + to_string( type->difficulty ) );
```
Comment (Night-Pryanik) [inline src/monster.cpp]:
```suggestion
        wprintz( w, c_light_gray, _( " Difficulty " ) + to_string( type->difficulty ) );
```

File: src/monster.cpp [hunk 2]
```diff
@@ -567,8 +571,27 @@ std::string monster::extended_description() const
     std::ostringstream ss;
     const auto att = get_attitude();
     std::string att_colored = colorize( att.first, att.second );
+    std::string difficulty_str = "";
```
Comment (Night-Pryanik) [inline src/monster.cpp]:
```suggestion
    std::string difficulty_str;
```

File: src/monster.cpp [hunk 3]
```diff
@@ -567,8 +571,27 @@ std::string monster::extended_description() const
...
+            difficulty_str = _( "<color_red>Very dangerous.</color>" );
+        } else if( type->difficulty < 50 ) {
+            difficulty_str = _( "<color_red>Extremely dangerous.</color>" );
+        } else {
+            difficulty_str = _( "<color_red>Fatally dangerous!</color>" );
+        }
+    }
 
-    ss << string_format( _( "This is a %s. %s" ), name().c_str(), att_colored.c_str() ) << std::endl;
+    ss << string_format( _( "This is a %s. %s  %s" ), name(), att_colored,
```
Comment (Night-Pryanik) [inline src/monster.cpp]:
```suggestion
    ss << string_format( _( "This is a %s.  %s %s" ), name(), att_colored,
```

**Comment by I-am-Erk** [comment]
Can you give a few illustrative examples of difficulties as calculated with this formula? Maybe, like, a standard zombie, and then a few creatures that have special abilities but aren't so tactical as the necromancer, like acid and corrosives?

**Comment by mlangsdorf** [comment]
Selected monsters, including the old JSON value, the new calculated value, my rough assessment of the danger, and all the factors that go into calculating it:


Name | json_diff | calc_diff | melee | melee_damage | dodge | armor | specials | fields | hp | speed | cost | agro | morale | vision_day | vision_night | Threat | fudge
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
shoggoth | 20 | 165 | 10 | 77 | 0 | 40 | 1 | 0 | 400 | 90 | 100 | 100 | 100 | 40 | 1 | High | 0
Beagle Mini-Tank UGV | 52 | 155 | 5 | 25 | 0 | 180 | 1 | 0 | 240 | 75 | 100 | 100 | 100 | 50 | 1 | High | 40
wraith | 50 | 140 | 6 | 33 | 3 | 20 | 2 | 0 | 800 | 140 | 100 | 2 | 500 | 50 | 1 | High | 0
jabberwock | 50 | 123 | 10 | 44 | 3 | 20 | 1 | 0 | 400 | 140 | 100 | 100 | 100 | 50 | 3 | High | 0
skeletal juggernaut | 50 | 112 | 5 | 63 | 0 | 65 | 4 | 0 | 480 | 60 | 100 | 100 | 100 | 50 | 4 | High | 0
flying polyp | 42 | 108 | 8 | 42 | 7 | 7 | 0 | 0 | 350 | 280 | 100 | 100 | 100 | 40 | 1 | High | 0
fungal tower | 40 | 100 | 9 | 27 | 3 | 35 | 1 | 0 | 300 | 100 | 100 | 100 | 100 | 40 | 1 | High | 10
flaming eye | 27 | 83 | 0 | 0 | 1 | 3 | 1 | 0 | 300 | 90 | 100 | 0 | 100 | 50 | 40 | High | 20
alpha razorclaw | 50 | 81 | 6 | 100 | 2 | 50 | 2 | 0 | 120 | 95 | 100 | 15 | 100 | 30 | 20 | High | 10
giant fungal blossom | 40 | 79 | 0 | 0 | 0 | 20 | 1 | 0 | 300 | 100 | 100 | 100 | 100 | 40 | 1 | High | 20
shocker brute | 22 | 71 | 3 | 18 | 0 | 11 | 2 | 0 | 275 | 110 | 100 | 100 | 100 | 40 | 3 | High | 15
zombie hulk | 50 | 67 | 5 | 32 | 0 | 20 | 1 | 0 | 480 | 130 | 100 | 100 | 100 | 50 | 4 | High | 2
mi-go | 26 | 60 | 7 | 48 | 4 | 16 | 2 | 0 | 210 | 120 | 100 | 20 | 30 | 50 | 1 | High | 0
zombie master | 16 | 59 | 5 | 12 | 4 | 4 | 1 | 0 | 180 | 90 | 100 | 5 | 100 | 50 | 5 | High | 20
mininuke hack | 40 | 54 | 0 | 0 | 0 | 7 | 1 | 0 | 20 | 150 | 100 | 100 | 100 | 40 | 1 | High | 40
zombie elite grenadier | 40 | 53 | 5 | 12 | 1 | 20 | 1 | 0 | 100 | 100 | 100 | 5 | 100 | 35 | 3 | High | 30
feral predator | 50 | 51 | 5 | 36 | 5 | 10 | 3 | 0 | 90 | 140 | 100 | 100 | 100 | 45 | 15 | High | 10
chicken walker | 32 | 50 | 3 | 9 | 0 | 32 | 1 | 0 | 90 | 115 | 100 | 100 | 100 | 50 | 35 | High | 20
zombie necromancer | 16 | 42 | 4 | 24 | 2 | -2 | 1 | 0 | 100 | 100 | 100 | 5 | 100 | 50 | 5 | High | 20
triffid queen | 28 | 40 | 7 | 32 | 0 | 20 | 1 | 0 | 280 | 85 | 100 | 100 | 200 | 40 | 1 | High | 0
homunculus | 13 | 37 | 8 | 48 | 6 | 6 | 0 | 0 | 120 | 110 | 100 | 100 | 100 | 40 | 1 | High | 0
zombear | 12 | 35 | 6 | 48 | 2 | 4 | 0 | 0 | 180 | 120 | 100 | 50 | 100 | 30 | 5 | High | 0
zombie nightstalker | 24 | 35 | 5 | 32 | 4 | 13 | 3 | 0 | 200 | 90 | 100 | 100 | 100 | 5 | 40 | High | 0
brain blob | 40 | 34 | 5 | 8 | 0 | 11 | 1 | 0 | 400 | 50 | 100 | 50 | 100 | 40 | 1 | High | 5
tribot | 26 | 33 | 10 | 54 | 0 | 20 | 1 | 0 | 80 | 90 | 100 | 100 | 100 | 40 | 1 | High | 0
antimateriel turret | 30 | 32 | 0 | 0 | 0 | 30 | 1 | 0 | 40 | 100 | 100 | 100 | 100 | 40 | 1 | High | 30
corrosive zombie | 25 | 32 | 1 | 6 | 0 | 12 | 2 | 0 | 140 | 75 | 100 | 100 | 100 | 40 | 3 | High | 15
police bot | 12 | 30 | 4 | 6 | 8 | 20 | 1 | 0 | 80 | 100 | 100 | 100 | 100 | 50 | 1 | High | 10
riot control bot | 11 | 29 | 0 | 0 | 8 | 20 | 1 | 0 | 80 | 100 | 100 | 100 | 100 | 50 | 1 | Medium | 10
gigantic naked mole-rat | 15 | 28 | 8 | 64 | 1 | 7 | 4 | 0 | 120 | 80 | 100 | -5 | 40 | 0 | 0 | Medium | 0
incandescent husk | 50 | 28 | 5 | 2 | 0 | -2 | 0 | 1 | 200 | 50 | 100 | 100 | 100 | 40 | 3 | Medium | 5
armored zombie | 25 | 27 | 5 | 18 | 0 | 128 | 0 | 0 | 120 | 70 | 100 | 100 | 100 | 30 | 3 | Medium | 10
security bot | 19 | 26 | 0 | 0 | 0 | 28 | 1 | 0 | 80 | 70 | 100 | 100 | 100 | 50 | 1 | Medium | 20
feral hunter | 6 | 25 | 4 | 12 | 3 | -2 | 3 | 0 | 90 | 110 | 100 | 100 | 100 | 40 | 5 | Medium | 10
shocker zombie | 12 | 25 | 4 | 3 | 2 | -2 | 1 | 0 | 85 | 105 | 100 | 100 | 100 | 40 | 1 | Medium | 15
acidic soldier ant | 21 | 24 | 7 | 16 | 2 | 26 | 1 | 0 | 100 | 100 | 100 | 40 | 100 | 5 | 5 | Medium | 10
razorclaw | 20 | 24 | 6 | 52 | 2 | 28 | 2 | 0 | 80 | 85 | 100 | 4 | 100 | 30 | 15 | Medium | 0
survivor zombie | 8 | 24 | 6 | 24 | 3 | 14 | 4 | 0 | 120 | 90 | 100 | 75 | 100 | 50 | 3 | Medium | 0
black bear | 10 | 23 | 6 | 40 | 3 | 1 | 1 | 0 | 100 | 140 | 100 | -10 | 60 | 30 | 10 | Medium | 0
screecher zombie | 15 | 23 | 5 | 27 | 2 | -2 | 3 | 0 | 85 | 100 | 100 | 100 | 100 | 50 | 8 | Medium | 5
milspec turret | 20 | 22 | 0 | 0 | 0 | 30 | 1 | 0 | 40 | 100 | 100 | 100 | 100 | 40 | 1 | Medium | 20
zombie bio-operator | 25 | 22 | 7 | 18 | 3 | 28 | 1 | 0 | 120 | 95 | 100 | 100 | 100 | 50 | 3 | Medium | 0
feral runner | 4 | 21 | 4 | 16 | 1 | -2 | 2 | 0 | 70 | 105 | 100 | 100 | 100 | 40 | 3 | Medium | 10
irradiated wanderer+++ | 20 | 21 | 8 | 22 | 2 | 16 | 1 | 0 | 120 | 100 | 100 | 100 | 100 | 50 | 3 | Medium | 0
laser turret | 14 | 20 | 0 | 0 | 0 | 30 | 1 | 0 | 30 | 100 | 100 | 100 | 100 | 40 | 1 | Medium | 20
zombie brute | 20 | 19 | 4 | 30 | 0 | 10 | 2 | 0 | 120 | 105 | 100 | 100 | 100 | 40 | 4 | Low | 2
smoker zombie | 10 | 18 | 4 | 8 | 4 | -2 | 2 | 1 | 75 | 110 | 100 | 100 | 100 | 50 | 3 | Low | 0
cougar | 12 | 16 | 6 | 20 | 5 | 0 | 2 | 0 | 60 | 180 | 100 | 0 | 40 | 40 | 10 | Low | 0
gracken | 15 | 16 | 10 | 16 | 4 | -2 | 0 | 0 | 100 | 180 | 100 | 0 | 100 | 40 | 1 | Low | 0
acidic zombie | 5 | 13 | 3 | 6 | 1 | -2 | 1 | 0 | 65 | 80 | 100 | 100 | 100 | 40 | 3 | Low | 10
moose | 1 | 13 | 6 | 12 | 1 | 5 | 1 | 0 | 120 | 200 | 100 | 5 | 80 | 40 | 7 | Low | 0
giant wasp | 20 | 12 | 5 | 28 | 4 | 12 | 0 | 0 | 40 | 150 | 100 | 10 | 40 | 10 | 5 | Low | 2
turret | 14 | 12 | 0 | 0 | 0 | 30 | 2 | 0 | 30 | 100 | 100 | 100 | 100 | 40 | 1 | Low | 10
antlered horror | 14 | 11 | 6 | 12 | 0 | 10 | 0 | 0 | 210 | 140 | 100 | 50 | 100 | 30 | 5 | Low |  
grappler zombie | 15 | 11 | 5 | 20 | 2 | -2 | 3 | 0 | 110 | 85 | 100 | 100 | 100 | 30 | 5 | Low | 0
zombie scientist | 3 | 11 | 3 | 12 | 1 | -2 | 2 | 0 | 80 | 75 | 100 | 100 | 100 | 50 | 3 | Low | 5
spitter zombie | 9 | 10 | 3 | 12 | 1 | -2 | 1 | 0 | 70 | 95 | 100 | 100 | 100 | 40 | 3 | Low | 5
cattle dog | 10 | 8 | 4 | 16 | 4 | -2 | 2 | 0 | 20 | 200 | 100 | -10 | 10 | 40 | 6 |   | 0
eyebot | 2 | 8 | 0 | 0 | 3 | 18 | 1 | 0 | 20 | 120 | 100 | 0 | 100 | 40 | 1 |   | 5
tough zombie | 3 | 8 | 4 | 12 | 1 | 3 | 3 | 0 | 100 | 85 | 100 | 100 | 100 | 40 | 3 |   | 0
soldier ant | 16 | 7 | 7 | 14 | 2 | 22 | 0 | 0 | 80 | 100 | 100 | 30 | 100 | 5 | 5 |   | 0
swimmer zombie | 10 | 7 | 3 | 16 | 2 | -2 | 3 | 0 | 70 | 95 | 100 | 100 | 100 | 40 | 3 |   | 0
wolf | 12 | 7 | 7 | 14 | 4 | 0 | 0 | 0 | 40 | 165 | 100 | 0 | 20 | 40 | 5 |   | 0
giant cellar spider | 20 | 6 | 6 | 12 | 7 | 3 | 0 | 0 | 45 | 150 | 100 | -10 | 100 | 5 | 5 |   | 0
rottweiler | 10 | 5 | 4 | 16 | 3 | -2 | 1 | 0 | 42 | 135 | 100 | 4 | 8 | 40 | 6 |   | 0
skeletal dog | 12 | 5 | 4 | 15 | 3 | 14 | 1 | 0 | 12 | 100 | 100 | 100 | 100 | 40 | 3 |   | 0
skitterbot | 12 | 5 | 0 | 0 | 0 | 24 | 1 | 0 | 40 | 130 | 100 | 100 | 100 | 50 | 1 |   | 2
fat zombie | 3 | 4 | 3 | 8 | 0 | 8 | 2 | 0 | 95 | 55 | 100 | 100 | 100 | 40 | 3 |   | 0
firefighter zombie | 3 | 4 | 5 | 9 | 0 | 12 | 1 | 0 | 80 | 70 | 100 | 100 | 100 | 40 | 3 |   | 0
zombie | 3 | 4 | 4 | 6 | 0 | -2 | 3 | 0 | 80 | 70 | 100 | 100 | 100 | 40 | 3 |   | 0
anklebiter | 3 | 3 | 4 | 12 | 3 | -2 | 1 | 0 | 35 | 105 | 100 | 100 | 100 | 30 | 3 |   | 0
kreck | 6 | 3 | 6 | 5 | 8 | 4 | 0 | 0 | 30 | 100 | 100 | 100 | 100 | 50 | 1 |   | 0
giant ant | 7 | 2 | 5 | 12 | 1 | 15 | 1 | 0 | 40 | 100 | 100 | 0 | 60 | 5 | 5 |   | 0
fungal zombie | 6 | 1 | 3 | 6 | 0 | 2 | 2 | 0 | 60 | 45 | 100 | 100 | 100 | 5 | 5 |   | 0
rattlesnake | 2 | 1 | 5 | 8 | 1 | -2 | 1 | 0 | 12 | 90 | 100 | -50 | 60 | 30 | 5 |   | 2
ant larva | 0 | -1 | 2 | 3 | 0 | -2 | 0 | 0 | 10 | 15 | 100 | -1 | 10 | 40 | 1 |   | 0
yellow chick | 10 | -1 | 1 | 2 | 1 | -2 | 0 | 0 | 1 | 40 | 100 | -99 | -8 | 40 | 1 |   | 0
dermatik larva | 4 | -2 | 1 | 2 | 1 | -2 | 1 | 0 | 10 | 20 | 100 | -20 | -20 | 40 | 1 |   | 2

**Comment by mlangsdorf** [comment]
Upped the spitter's diff to 10 and the corrosive to 15, and adjusted the docs to match.  A spitter is now difficulty 16, an acid zombie difficulty 13, and a corrosive is difficulty 32, which feels about right to me - spitters aren't much worse than acid zombies, but corrosive zombies are a whole lot worse.

**Pull request merged**